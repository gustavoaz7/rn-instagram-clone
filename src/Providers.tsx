import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import { theme as stylesTheme } from './styles/theme';
import { store as reduxStore } from './redux/store';

type TProvidersProps = {
  children: ReactNode;
  store?: typeof reduxStore;
  theme?: typeof stylesTheme;
};

export function Providers({
  children,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>
  );
}
