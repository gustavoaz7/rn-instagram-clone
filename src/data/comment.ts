import { TComment } from '../types';

export const generateComment = (): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
});
