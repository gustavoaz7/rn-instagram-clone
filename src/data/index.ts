import type {
  TLike,
  TComment,
  TPostMedia,
  TPost,
  TOwner,
  TStory,
  TStoryMedia,
  TProfile,
  TTappableObject,
} from '../types';

export const generateMockOwner = (): TOwner => ({
  id: `${Math.random()}`,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
});

export const generateMockLike = (): TLike => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
});

export const generateMockComment = (likesQty = 1): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  viewerHasLiked: false,
});

export const generateMockTappableObject = ({
  type = 'mention',
  text = 'random',
  x = 0.1,
  y = 0.2,
  scale = 1.1,
  rotation = 300,
}: Partial<TTappableObject> = {}): TTappableObject => ({
  id: `${Math.random()}`,
  type,
  text,
  x,
  y,
  scale,
  rotation,
});

export const generateMockPostMedia = ({ index = 0 } = {}): TPostMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
});

export const generateMockPost = ({
  mediasQty = 1,
  likesQty = 3,
  commentsQty = 1,
} = {}): TPost => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
  medias: [...Array(mediasQty)].map(generateMockPostMedia),
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  previewComments: {
    count: commentsQty,
    comments: [...Array(commentsQty)].map(generateMockComment),
  },
  location: 'Daronville',
  viewerHasLiked: false,
});

export const generateMockStoryMedia = ({
  viewersQty = 0,
  index = 0,
} = {}): TStoryMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
  previewViewers: {
    count: viewersQty,
    viewers: [...Array(viewersQty)].map(generateMockOwner),
  },
  takenAt: 1620488304326,
  expiresAt: 1620488304326 + 3600 * 24,
});

export const generateMockStory = ({
  mediasQty = 3,
  viewersQty = 0,
} = {}): TStory => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  expiresAt: 0,
  latestMediaAt: 0,
  seenAt: 0,
  medias: [...Array(mediasQty)].map((_, index) =>
    generateMockStoryMedia({ viewersQty, index }),
  ),
});

export const generateMockProfile = (): TProfile => ({
  bio: 'animi eum earum aut explicabo',
  followCount: 47,
  followedByCount: 50,
  followedByViewer: false,
  followsViewer: false,
  fullName: 'Gustavo Az',
  postsCount: 20,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
  story: generateMockStory(),
});
