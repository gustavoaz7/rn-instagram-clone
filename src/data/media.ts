import type { TPostMedia } from '../types';

export const generatePostMedia = (): TPostMedia => ({
  id: `${Math.random()}`,
  url: 'http://placeimg.com/640/480/cats',
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  tappableObjects: [],
});
