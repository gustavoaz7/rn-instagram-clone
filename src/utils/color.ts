const HEX_COLOR_RE = /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/;

export function addAlphaToHEX(hex: string, alpha: number) {
  if (!HEX_COLOR_RE.test(hex) || alpha < 0 || alpha > 1) return hex;

  let cleanHex = hex;
  if (hex.length === 4) {
    cleanHex = `#${hex[1].repeat(2)}${hex[2].repeat(2)}${hex[3].repeat(2)}`;
  }

  return `${cleanHex}${Math.round(255 * alpha).toString(16)}`;
}
