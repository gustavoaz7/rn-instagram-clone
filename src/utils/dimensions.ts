import { Dimensions } from 'react-native';

export const SCREEN_WIDTH = Dimensions.get('screen').width;
