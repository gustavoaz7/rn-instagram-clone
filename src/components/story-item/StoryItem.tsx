import React, { useCallback, useEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components/native';
import { GestureResponderEvent, Animated } from 'react-native';
import DirectSvg from '../../../assets/svg/direct.svg';
import MenuVerticalSvg from '../../../assets/svg/menu-vertical.svg';
import { SCREEN_HEIGHT, SCREEN_WIDTH } from '../../utils/dimensions';
import { TStory } from '../../types';
import { Text } from '../text';
import { dateToString } from '../../utils/date';

type TStoryItemProps = {
  story: TStory;
  storyIndex: number;
  isCurrentStory: boolean;
  initialMediaIndex?: number;
  shouldPauseAnimations?: boolean;
  goToStory: (nextStoryIndex: number) => void;
};

export const StoryItem = ({
  story,
  storyIndex,
  isCurrentStory,
  initialMediaIndex = 0,
  shouldPauseAnimations = false,
  goToStory,
}: TStoryItemProps) => {
  const [mediaIndex, setMediaIndex] = useState(initialMediaIndex);
  const [isAnimationPaused, setIsAnimationPaused] = useState(false);
  const [text, setText] = useState('');
  const handleTextChange = useCallback(newText => {
    setText(newText);
  }, []);

  const progressBars = useRef<Animated.Value[]>(
    story.medias.map(() => new Animated.Value(0)),
  ).current;

  const stopProgressAnimation = useCallback(() => {
    progressBars[mediaIndex].stopAnimation();
  }, [progressBars, mediaIndex]);

  const goToPrevMedia = useCallback(() => {
    progressBars[mediaIndex].setValue(0);
    if (mediaIndex > 0) {
      progressBars[mediaIndex - 1].setValue(0);
      setMediaIndex(mediaIndex - 1);
    } else {
      goToStory(storyIndex - 1);
    }
  }, [mediaIndex, goToStory, storyIndex, progressBars]);

  const goToNextMedia = useCallback(() => {
    if (mediaIndex < story.medias.length - 1) {
      setMediaIndex(mediaIndex + 1);
    } else {
      const bar = progressBars[mediaIndex];
      bar.setValue(0);
      goToStory(storyIndex + 1);
    }
  }, [mediaIndex, goToStory, storyIndex, story.medias.length, progressBars]);

  const handlePress = useCallback(
    (e: GestureResponderEvent) => {
      stopProgressAnimation();
      const bar = progressBars[mediaIndex];
      if (e.nativeEvent.locationX < SCREEN_WIDTH / 2) {
        bar.setValue(0);
        goToPrevMedia();
      } else {
        bar.setValue(1);
        goToNextMedia();
      }
    },
    [
      stopProgressAnimation,
      progressBars,
      mediaIndex,
      goToPrevMedia,
      goToNextMedia,
    ],
  );

  const startProgressAnimation = useCallback(() => {
    const bar = progressBars[mediaIndex];
    Animated.timing(bar, {
      toValue: 1,
      // Adjusts duration when re-starting after stop animation
      // eslint-disable-next-line
      duration: 5000 - (bar as any).__getValue() * 5000,
      useNativeDriver: false,
    }).start(({ finished }) => {
      if (finished) {
        goToNextMedia();
      }
    });
  }, [mediaIndex, progressBars, goToNextMedia]);

  useEffect(() => {
    if (isCurrentStory && !shouldPauseAnimations) {
      startProgressAnimation();
    } else {
      if (!shouldPauseAnimations) {
        progressBars[mediaIndex].setValue(0);
      }
      stopProgressAnimation();
    }
  }, [
    mediaIndex,
    isCurrentStory,
    progressBars,
    shouldPauseAnimations,
    startProgressAnimation,
    stopProgressAnimation,
  ]);

  useEffect(() => {
    if (shouldPauseAnimations || isAnimationPaused) {
      stopProgressAnimation();
    }
  }, [shouldPauseAnimations, isAnimationPaused, stopProgressAnimation]);

  const handleLongPress = useCallback(() => {
    setIsAnimationPaused(true);
  }, [setIsAnimationPaused]);

  const handlePressOut = useCallback(() => {
    if (isAnimationPaused) {
      startProgressAnimation();
    }
    setIsAnimationPaused(false);
  }, [setIsAnimationPaused, isAnimationPaused, startProgressAnimation]);

  const currentMedia = story.medias[mediaIndex];

  return (
    <Container
      onPress={handlePress}
      onLongPress={handleLongPress}
      onPressOut={handlePressOut}
      delayLongPress={200}
    >
      <Image source={{ uri: currentMedia.url }} />
      <ProgressBarContainer>
        {story.medias.map((_, i) => (
          <ProgressBarItem
            // eslint-disable-next-line react/no-array-index-key
            key={`bar-${i}`}
          >
            <ProgressBarItemAnimated
              style={{
                width: progressBars[i].interpolate({
                  inputRange: [0, 1],
                  outputRange: ['0%', '100%'],
                  extrapolate: 'clamp',
                }),
              }}
            />
          </ProgressBarItem>
        ))}
      </ProgressBarContainer>
      <Header>
        <AvatarContainer>
          <Avatar source={{ uri: story.owner.profilePicUrl }} />
          <Username>{story.owner.username}</Username>
          <Time>{dateToString(new Date(currentMedia.takenAt), true)}</Time>
        </AvatarContainer>
        <MenuVerticalIcon />
      </Header>
      <Footer>
        <Input
          value={text}
          onChangeText={handleTextChange}
          onFocus={stopProgressAnimation}
          onBlur={startProgressAnimation}
        />
        <DirectIcon />
      </Footer>
    </Container>
  );
};

const Container = styled.Pressable`
  flex: 1;
  justify-content: space-between;
  width: ${SCREEN_WIDTH}px;
`;

const CenteredRow = styled.View`
  align-items: center;
  flex-direction: row;
`;

const Header = styled(CenteredRow)`
  padding-top: ${({ theme }) => theme.spacing.s};
  justify-content: space-between;
`;

const Image = styled.Image`
  position: absolute;
  top: 0;
  left: 0;
  width: ${SCREEN_WIDTH}px;
  height: ${SCREEN_HEIGHT}px;
  border-radius: 5px;
`;

const AvatarContainer = styled(CenteredRow)`
  padding: ${({ theme }) => theme.spacing.m};
`;

const Avatar = styled.Image`
  width: 36px;
  height: 36px;
  border-radius: 36px;
  margin-right: ${({ theme }) => theme.spacing.s};
`;

const Username = styled(Text)`
  font-weight: bold;
  color: ${({ theme }) => theme.color.white};
`;

const Time = styled(Text)`
  color: ${({ theme }) => theme.color.white}CC;
  margin-left: ${({ theme }) => theme.spacing.s};
`;

const MenuVerticalIcon = styled(MenuVerticalSvg).attrs(({ theme }) => ({
  color: theme.color.white,
}))`
  margin-right: ${({ theme }) => theme.spacing.m};
`;

const Footer = styled(CenteredRow)`
  padding: ${({ theme }) => `${theme.spacing.s} ${theme.spacing.m}`};
`;

const Input = styled.TextInput.attrs(({ theme }) => ({
  placeholder: 'Send message',
  placeholderTextColor: theme.color.white,
  multiline: true,
}))`
  flex: 1;
  border-radius: 25px;
  padding: ${({ theme }) => `${theme.spacing.s} ${theme.spacing.m}`};
  max-height: 100px;
  border-width: 1px;
  border-color: ${({ theme }) => theme.color.white}80;
  color: ${({ theme }) => theme.color.white};
`;

const DirectIcon = styled(DirectSvg).attrs(({ theme }) => ({
  color: theme.color.white,
}))`
  margin-left: ${({ theme }) => theme.spacing.l};
`;

const ProgressBarContainer = styled(CenteredRow)`
  position: absolute;
  top: ${({ theme }) => theme.spacing.s};
  left: 3px;
  right: 3px;
  height: 2px;
  justify-content: space-between;
`;

const progressBarItemClass = css`
  border-radius: 2px;
  height: 100%;
  flex: 1;
`;

const ProgressBarItem = styled.View`
  ${progressBarItemClass};
  margin: 0 2px;
  background-color: ${({ theme }) => theme.color.white}80;
`;

const ProgressBarItemAnimated = styled(Animated.View)`
  ${progressBarItemClass};
  background-color: ${({ theme }) => theme.color.white};
`;
