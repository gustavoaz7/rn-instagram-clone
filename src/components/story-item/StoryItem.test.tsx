import React from 'react';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import { Animated, TextInput } from 'react-native';
import MockDate from 'mockdate';
import { StoryItem, STORY_TIMEOUT } from './StoryItem';
import { Providers } from '../../Providers';
import { generateMockStory } from '../../data';
import {
  setupTimeTravel,
  destroyTimeTravel,
  timeTravel,
} from '../../test/time-travel';
import { SCREEN_WIDTH } from '../../utils/dimensions';

const story = generateMockStory();

describe('components - StoryItem', () => {
  const options: RenderOptions = { wrapper: Providers };
  const defaultProps = {
    story,
    storyIndex: 0,
    isCurrentStory: true,
    goToStory: jest.fn(),
  };

  beforeEach(() => {
    setupTimeTravel();
    defaultProps.goToStory.mockClear();
  });

  afterEach(() => {
    destroyTimeTravel();
  });

  it('renders', () => {
    render(<StoryItem {...defaultProps} />, options);
  });

  it('matches snapshot', () => {
    MockDate.set(story.medias[0].takenAt + 3600 * 1000);
    const { toJSON } = render(<StoryItem {...defaultProps} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('starts all progress bars at 0', () => {
    const { getByTestId } = render(<StoryItem {...defaultProps} />, options);

    getByTestId('StoryItem-ProgressBar')
      .findAllByType(Animated.View)
      .forEach(progressBar => {
        expect(progressBar.props.style[1]).toMatchInlineSnapshot(`
          Object {
            "width": "0%",
          }
        `);
      });
  });

  describe('when it is not the current story', () => {
    it('does not animate progress bar', () => {
      const { getByTestId } = render(
        <StoryItem {...defaultProps} isCurrentStory={false} />,
        options,
      );

      const progressBarProps = getByTestId(
        'StoryItem-ProgressBar',
      ).findAllByType(Animated.View)[0].props;

      expect(progressBarProps.style[1]).toMatchInlineSnapshot(`
        Object {
          "width": "0%",
        }
      `);

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(progressBarProps.style[1]).toMatchInlineSnapshot(`
        Object {
          "width": "0%",
        }
      `);
    });

    it('does not change story image', () => {
      const { getByTestId } = render(
        <StoryItem {...defaultProps} isCurrentStory={false} />,
        options,
      );

      expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
        story.medias[0].url,
      );

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
        story.medias[0].url,
      );
    });
  });

  describe('when `shouldPauseAnimations` is true"', () => {
    it('pauses progress', () => {
      const { getByTestId, rerender } = render(
        <StoryItem {...defaultProps} />,
        options,
      );

      const progressBarProps = getByTestId(
        'StoryItem-ProgressBar',
      ).findAllByType(Animated.View)[0].props;

      act(() => {
        timeTravel(STORY_TIMEOUT / 2);
      });

      rerender(<StoryItem {...defaultProps} shouldPauseAnimations />);

      act(() => {
        timeTravel(STORY_TIMEOUT / 2);
      });

      // eslint-disable-next-line no-underscore-dangle
      expect((progressBarProps.style[1].width as any).__getValue()).toMatch(
        /50.\d+%/,
      );
    });

    it('does not change story image', () => {
      const { getByTestId } = render(
        <StoryItem {...defaultProps} shouldPauseAnimations />,
        options,
      );

      expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
        story.medias[0].url,
      );

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
        story.medias[0].url,
      );
    });
  });

  it('animates progress bar', () => {
    const { getByTestId } = render(<StoryItem {...defaultProps} />, options);

    const progressBarProps = getByTestId('StoryItem-ProgressBar').findAllByType(
      Animated.View,
    )[0].props;

    expect(progressBarProps.style[1]).toMatchInlineSnapshot(`
      Object {
        "width": "0%",
      }
    `);

    act(() => {
      timeTravel(STORY_TIMEOUT / 2);
    });

    expect(progressBarProps.style[1]).toMatchInlineSnapshot(`
      Object {
        "width": "50%",
      }
    `);

    act(() => {
      timeTravel(STORY_TIMEOUT / 2);
    });

    expect(progressBarProps.style[1]).toMatchInlineSnapshot(`
      Object {
        "width": "100%",
      }
    `);
  });

  describe(`when ${STORY_TIMEOUT}ms have ellapsed`, () => {
    it('changes story image', () => {
      const { getByTestId } = render(<StoryItem {...defaultProps} />, options);

      expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
        story.medias[0].url,
      );

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
        story.medias[1].url,
      );
    });

    it('animates next progress bar', () => {
      const { getByTestId } = render(<StoryItem {...defaultProps} />, options);

      const [prevProgressBar, currentProgressBar] = getByTestId(
        'StoryItem-ProgressBar',
      ).findAllByType(Animated.View);

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(prevProgressBar.props.style[1]).toMatchInlineSnapshot(`
        Object {
          "width": "100%",
        }
      `);

      expect(currentProgressBar.props.style[1]).toMatchInlineSnapshot(`
        Object {
          "width": "0%",
        }
      `);

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(currentProgressBar.props.style[1]).toMatchInlineSnapshot(`
        Object {
          "width": "100%",
        }
      `);
    });
  });

  it('starts correctly at provided `initialMediaIndex`', () => {
    const { getByTestId } = render(
      <StoryItem {...defaultProps} initialMediaIndex={1} />,
      options,
    );

    expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
      story.medias[1].url,
    );

    const [firstProgressBar, currentProgressBar] = getByTestId(
      'StoryItem-ProgressBar',
    ).findAllByType(Animated.View);

    expect(firstProgressBar.props.style[1]).toMatchInlineSnapshot(`
      Object {
        "width": "100%",
      }
    `);

    expect(currentProgressBar.props.style[1]).toMatchInlineSnapshot(`
      Object {
        "width": "0%",
      }
    `);
  });

  it('pauses progress on long press and resumes on press out', () => {
    const { getByTestId } = render(<StoryItem {...defaultProps} />, options);

    const container = getByTestId('StoryItem');
    const progressBarProps = getByTestId('StoryItem-ProgressBar').findAllByType(
      Animated.View,
    )[0].props;

    fireEvent(container, 'onLongPress');

    act(() => {
      timeTravel(STORY_TIMEOUT);
    });

    // eslint-disable-next-line no-underscore-dangle
    expect((progressBarProps.style[1].width as any).__getValue()).toMatch(
      /0.\d+%/,
    );
    expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
      story.medias[0].url,
    );

    fireEvent(container, 'onPressOut');

    act(() => {
      timeTravel(STORY_TIMEOUT);
    });

    // eslint-disable-next-line no-underscore-dangle
    expect((progressBarProps.style[1].width as any).__getValue()).toBe('100%');
    expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
      story.medias[1].url,
    );
  });

  it('pauses progress when input is focused and resumes on blur', () => {
    const { getByTestId, UNSAFE_getByType } = render(
      <StoryItem {...defaultProps} />,
      options,
    );

    const input = UNSAFE_getByType(TextInput);
    const progressBarProps = getByTestId('StoryItem-ProgressBar').findAllByType(
      Animated.View,
    )[0].props;

    fireEvent(input, 'onFocus');

    act(() => {
      timeTravel(STORY_TIMEOUT);
    });

    /* eslint-disable no-underscore-dangle */
    expect((progressBarProps.style[1].width as any).__getValue()).toMatch(
      /0.00\d+%/,
    );
    expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
      story.medias[0].url,
    );

    fireEvent(input, 'onBlur');

    act(() => {
      timeTravel(STORY_TIMEOUT);
    });

    expect((progressBarProps.style[1].width as any).__getValue()).toBe('100%');
    expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
      story.medias[1].url,
    );
    /* eslint-enable no-underscore-dangle */
  });

  describe('when at an intermediary media', () => {
    const mediaIndex = 1;

    describe('on left side press', () => {
      it('goes to previous media', () => {
        const { getByTestId } = render(
          <StoryItem {...defaultProps} initialMediaIndex={mediaIndex} />,
          options,
        );

        fireEvent(getByTestId('StoryItem'), 'onPress', {
          nativeEvent: { locationX: SCREEN_WIDTH / 2 - 1 },
        });

        expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
          story.medias[mediaIndex - 1].url,
        );
      });

      it('resets current progress bar and starts previous from 0', () => {
        const { getByTestId } = render(
          <StoryItem {...defaultProps} initialMediaIndex={mediaIndex} />,
          options,
        );

        fireEvent(getByTestId('StoryItem'), 'onPress', {
          nativeEvent: { locationX: SCREEN_WIDTH / 2 - 1 },
        });

        const [firstProgressBar, secondProgressBar] = getByTestId(
          'StoryItem-ProgressBar',
        ).findAllByType(Animated.View);

        expect(firstProgressBar.props.style[1]).toMatchInlineSnapshot(`
          Object {
            "width": "0%",
          }
        `);

        expect(secondProgressBar.props.style[1]).toMatchInlineSnapshot(`
          Object {
            "width": "0%",
          }
        `);
      });
    });

    describe('on right side press', () => {
      it('goes to previous media', () => {
        const { getByTestId } = render(
          <StoryItem {...defaultProps} initialMediaIndex={mediaIndex} />,
          options,
        );

        fireEvent(getByTestId('StoryItem'), 'onPress', {
          nativeEvent: { locationX: SCREEN_WIDTH / 2 + 1 },
        });

        expect(getByTestId('StoryItem-Image').props.source.uri).toBe(
          story.medias[mediaIndex + 1].url,
        );
      });

      it('completes current progress bar and starts next from 0', () => {
        const { getByTestId } = render(
          <StoryItem {...defaultProps} initialMediaIndex={mediaIndex} />,
          options,
        );

        fireEvent(getByTestId('StoryItem'), 'onPress', {
          nativeEvent: { locationX: SCREEN_WIDTH / 2 + 1 },
        });

        const [, secondProgressBar, thirdProgressBar] = getByTestId(
          'StoryItem-ProgressBar',
        ).findAllByType(Animated.View);

        expect(secondProgressBar.props.style[1]).toMatchInlineSnapshot(`
          Object {
            "width": "100%",
          }
        `);

        expect(thirdProgressBar.props.style[1]).toMatchInlineSnapshot(`
          Object {
            "width": "0%",
          }
        `);
      });
    });
  });

  describe('when at last media', () => {
    const storyIndex = Math.random();

    it(`calls "goToStory" with next storyIndex after ${STORY_TIMEOUT}ms`, () => {
      render(
        <StoryItem
          {...defaultProps}
          initialMediaIndex={story.medias.length - 1}
          storyIndex={storyIndex}
        />,
        options,
      );

      act(() => {
        timeTravel(STORY_TIMEOUT);
      });

      expect(defaultProps.goToStory).toHaveBeenCalledTimes(1);
      expect(defaultProps.goToStory).toHaveBeenCalledWith(storyIndex + 1);
    });

    it(`calls "goToStory" with next storyIndex on right press`, () => {
      const { getByTestId } = render(
        <StoryItem
          {...defaultProps}
          initialMediaIndex={story.medias.length - 1}
          storyIndex={storyIndex}
        />,
        options,
      );

      fireEvent(getByTestId('StoryItem'), 'onPress', {
        nativeEvent: { locationX: SCREEN_WIDTH / 2 + 1 },
      });

      expect(defaultProps.goToStory).toHaveBeenCalledTimes(1);
      expect(defaultProps.goToStory).toHaveBeenCalledWith(storyIndex + 1);
    });
  });

  describe('when at first media', () => {
    const storyIndex = Math.random();

    it(`calls "goToStory" with previous storyIndex on left side press`, () => {
      const { getByTestId } = render(
        <StoryItem {...defaultProps} storyIndex={storyIndex} />,
        options,
      );

      fireEvent(getByTestId('StoryItem'), 'onPress', {
        nativeEvent: { locationX: SCREEN_WIDTH / 2 - 1 },
      });

      expect(defaultProps.goToStory).toHaveBeenCalledTimes(1);
      expect(defaultProps.goToStory).toHaveBeenCalledWith(storyIndex - 1);
    });
  });
});
