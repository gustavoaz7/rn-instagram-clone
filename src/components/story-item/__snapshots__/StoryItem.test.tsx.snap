// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - StoryItem matches snapshot 1`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "space-between",
          "width": 750,
        },
      ]
    }
    testID="StoryItem"
  >
    <Image
      source={
        Object {
          "uri": "http://placeimg.com/640/480/cats?0",
        }
      }
      style={
        Array [
          Object {
            "borderRadius": 5,
            "height": 1334,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 750,
          },
        ]
      }
      testID="StoryItem-Image"
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 2,
            "justifyContent": "space-between",
            "left": 3,
            "position": "absolute",
            "right": 3,
            "top": 10,
          },
        ]
      }
      testID="StoryItem-ProgressBar"
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF80",
              "borderRadius": 2,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "marginBottom": 0,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 2,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "width": "0%",
            }
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF80",
              "borderRadius": 2,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "marginBottom": 0,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 2,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "width": "0%",
            }
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF80",
              "borderRadius": 2,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "marginBottom": 0,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 2,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": "100%",
              "width": "0%",
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingTop": 10,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingBottom": 15,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 15,
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "uri": "https://avatars.githubusercontent.com/u/28453820?v=4",
            }
          }
          style={
            Array [
              Object {
                "borderRadius": 36,
                "height": 36,
                "marginRight": 10,
                "width": 36,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#262626",
              },
              Object {
                "color": "#FFFFFF",
                "fontWeight": "bold",
              },
            ]
          }
        >
          gustavoaz7
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#262626",
              },
              Object {
                "color": "#FFFFFFCC",
                "marginLeft": 10,
              },
            ]
          }
        >
          1 h
        </Text>
      </View>
      <SvgMock
        color="#FFFFFF"
        style={
          Array [
            Object {
              "marginRight": 15,
            },
          ]
        }
      />
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "left": 0,
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
          "position": "absolute",
          "right": 0,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF80",
            "borderRadius": 25,
            "borderWidth": 1,
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "paddingBottom": 10,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 10,
          },
        ]
      }
      testID="StoryItem-FakeInput"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "color": "#FFFFFF",
            },
          ]
        }
      >
        Send message
      </Text>
    </View>
    <SvgMock
      color="#FFFFFF"
      style={
        Array [
          Object {
            "marginLeft": 20,
          },
        ]
      }
    />
  </View>,
]
`;
